src/languages/latex.js

```diff
@@ -1,70 +1,231 @@
-      {
-        className: 'formula',
-        contains: [COMMAND],
+  var ACTIVE_CHAR = {
+    className: 'keyword',
+    relevance: 0,
+    begin: /~|"/
+  };
+  var MACRO_PARAM = {
+    className: 'params',
+    relevance: 0,
+    begin: /#+\d?/
+  };
+  var DOUBLE_CARET_CHAR = {
+    variants: DOUBLE_CARET_VARIANTS
+  };
+  var SPECIAL_CATCODE = {
+    className: 'built_in',
+    relevance: 0,
+    begin: /[$&^_]/
+  };
+  var BRACES = {
+    className: 'built_in',
+    relevance: 0,
+    begin: /[{}]/
+  };
+  var MAGIC_COMMENT = hljs.COMMENT(
+    '% !TeX',
+    '$',
+    {
+      className: 'meta',
+      relevance: 10
+    }
+  );
+  var COMMENT = hljs.COMMENT(
+    '%',
+    '$',
+    {
+      relevance: 0
+    }
+  );
+  var EVERYTHING_BUT_BRACES_AND_VERBATIM = [
+    CONTROL_SEQUENCE,
+    ACTIVE_CHAR,
+    MACRO_PARAM,
+    DOUBLE_CARET_CHAR,
+    SPECIAL_CATCODE,
+    MAGIC_COMMENT,
+    COMMENT
+  ];
+  var EVERYTHING_BUT_VERBATIM = [...EVERYTHING_BUT_BRACES_AND_VERBATIM, BRACES];
+  var BRACE_GROUP_NO_VERBATIM = {
+    begin: /\{/, end: /\}/,
+    keywords: {
+      $pattern: /\{|\}/,
+      built_in: '{ }'
+    },
+    relevance: 0,
+    contains: ['self', ...EVERYTHING_BUT_BRACES_AND_VERBATIM]
+  }
+  var ARGUMENT_BRACES = hljs.inherit(
+    BRACE_GROUP_NO_VERBATIM,
+    {endsParent: true, contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_BRACES_AND_VERBATIM]}
+  );
+  var ARGUMENT_BRACKETS = {
+    begin: /\[/, end: /\]/,
+    endsParent: true,
+    relevance: 0,
+    contains: [BRACE_GROUP_NO_VERBATIM, ...EVERYTHING_BUT_BRACES_AND_VERBATIM]
+  }
+  var ARGUMENT_ABSENT = {
+    begin: /(?=[.$])/, end: /(?=[.$])/,
+    endsParent: true,
+    relevance: 0
+  }
+  var SPACE_GOBBLER = {
+    begin: /\s+/,
+    endsParent: true,
+    relevance: 0
+  };
+  var ARGUMENT_M = [ARGUMENT_BRACES];
+  var ARGUMENT_O = [ARGUMENT_BRACKETS, ARGUMENT_ABSENT];
+  var ARGUMENT_NONE = [ARGUMENT_ABSENT];
+  var ARGUMENT_AND_THEN = function(arg, starts_mode) {
+    return {
+      contains: [SPACE_GOBBLER],
+      starts: {
         relevance: 0,
-        variants: [
-          {begin: /\$\$/, end: /\$\$/},
-          {begin: /\$/, end: /\$/}
-        ]
+        contains: arg,
+        starts: starts_mode
+      }
+    };
+  };
+  var CSNAME = function(csname, starts_mode) {
+    return hljs.inherit(
+      {
+        begin: '\\\\' + csname + '(?![a-zA-Z@:_])',
+        keywords: {$pattern: /\S+/, keyword: '\\' + csname}
       },
-      hljs.COMMENT(
-        '%',
-        '$',
-        {
-          relevance: 0
-        }
-      )
+      ARGUMENT_AND_THEN(ARGUMENT_NONE, starts_mode)
+    );
+  };
+  var BEGIN_ENV = function(envname, starts_mode) {
+    return hljs.inherit(
+      {
+        begin: '\\\\begin(?=\\s*\\r?\\n?\\s*\\{' + envname + '\\})',
+        keywords: {$pattern: /\S+/, keyword: '\\begin'}
+      },
+      ARGUMENT_AND_THEN(ARGUMENT_M, starts_mode)
+    );
+  };
+  var VERBATIM_DELIMITED_EQUAL = hljs.END_SAME_AS_BEGIN({
```

Comment:
Why do we need END_SAME_AS_BEGIN here, isn't this really just two different variants?

Reply:
Not sure what you mean here. (Maybe you mistook `.` for `\.` in the regex?) `VERBATIM_DELIMITED_EQUAL` looks for an arbitrary character (including newlines) and marks everything up to the next occurrence of that same character as verbatim (`string`).

Reply:
> Maybe you mistook . for \. in the regex?)

LOL.  That makes a lot more sense now. :)

Reply:
(The newline really is an edge case, but I thought including it won't hurt, as it doesn't change the rule otherwise.)